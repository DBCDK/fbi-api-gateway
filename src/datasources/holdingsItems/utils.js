// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
