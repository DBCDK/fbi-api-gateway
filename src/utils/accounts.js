/**
 * @file This file handles util function to select single or multiple CULR accounts
 *
 */

import { isValidCpr } from "./cpr";

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} Object or null
 */
export async function getAccount(accessToken, context, props) {
  return (await getAccounts(accessToken, context, props))?.[0] || null;
}

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} (Array)
 */
export async function getAccounts(accessToken, context, props) {
  // userInfo
  const userinfo = await context.datasources.getLoader("userinfo").load({
    accessToken,
  });

  const user = userinfo?.attributes;

  if (!user?.userId) {
    return [];
  }

  // select dataloader
  let dataloader = isValidCpr(user.userId)
    ? "culrGetAccountsByGlobalId"
    : "culrGetAccountsByLocalId";

  // Retrieve user culr account
  const response = await context.datasources.getLoader(dataloader).load({
    userId: user.userId,
    agencyId: user.loggedInAgencyId,
  });

  return filterAccountsByProps(response.accounts, props);
}

/**
 *
 * @param {Array} accounts
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAccountsByProps(accounts = [], props = {}) {
  if (props.agency) {
    accounts = accounts?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    accounts = accounts?.filter(({ userIdValue }) => userIdValue === props.id);
  }
  if (props.type) {
    accounts = accounts?.filter(({ userIdType }) => userIdType === props.type);
  }
  return accounts;
}

/**
 *
 * Function to select a specific agency from a userinfo agencies list
 *
 * @param {Array} agencies
 * @param {object} props - Filter agencies by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAgenciesByProps(agencies = [], props = {}) {
  if (props.agency) {
    agencies = agencies?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    agencies = agencies?.filter(({ userId }) => userId === props.id);
  }
  if (props.type) {
    agencies = agencies?.filter(({ userIdType }) => userIdType === props.type);
  }
  return agencies;
}
