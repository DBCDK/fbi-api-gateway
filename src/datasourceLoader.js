import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
