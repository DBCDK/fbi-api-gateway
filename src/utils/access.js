/**
 * @file access util functions
 */

import { filterDuplicateAgencies, resolveManifestation } from "./utils";

/**
 *
 * Check whether any of the users libraries has access to Infomedia.
 * The returned agencyId will be billed for infomedia access.
 *
 * First we check if the users login library has access to info media, if not,
 * we check if any other og the users associated accounts has access to infomedia,
 * if so, we return the first one found.
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
async function infomedia(context) {
  const user = context?.user;

  // get rights from idp
  const idpRights = await context.datasources.getLoader("idp").load("");

  // check if users loggedInAgency has infomedia access
  const loggedInAgencyId = user?.loggedInAgencyId;

  if (loggedInAgencyId && idpRights[loggedInAgencyId]) {
    return loggedInAgencyId;
  }

  /*
   * Alternativly check all users accounts
   */

  // filtered accounts
  const userInfoAccounts = filterDuplicateAgencies(user.agencies);

  const hasAccess = userInfoAccounts?.filter(
    ({ agencyId }) => idpRights[agencyId]
  );

  // check for infomedia access - if any of users agencies subscribes
  if (hasAccess.length > 0) {
    return hasAccess[0]?.agencyId;
  }

  return null;
}

/**
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
export async function getInfomediaAgencyId(context) {
  return await infomedia(context);
}

/**
 *
 * @param {object} context
 *
 * @returns {boolean}
 */
export async function hasInfomediaAccess(context) {
  return !!(await infomedia(context));
}

/**
 * Get type of access for manifestation with given pid.
 *  url access
 *  ill access
 *  infomedia access
 *  digital article service access
 *
 * @param pid
 * @param context
 * @returns {Promise<*>}
 */
export async function resolveAccess(manifestation, context) {
  // We parse the access structure from JED, and convert it
  // to the union type structure.
  // At some point we may choose to follow the structure of JED closely,
  // but it has to be coordinated with stakeholders

  let parent =
    typeof manifestation === "object"
      ? manifestation
      : await resolveManifestation({ pid: manifestation }, context);

  const res = [];

  // use linkchecker to check status of a single accessUrl
  // linkchecker does NOT handle proxy urls, so we skip ebookcentral and ebsco urls
  const linkStatus = async (url) => {
    if (url?.includes("ebookcentral") || url?.includes("ebscohost")) {
      return "OK";
    }

    const linkcheck = await context.datasources
      .getLoader("linkcheck")
      .load({ urls: [url] });
    const link = linkcheck.find((check) => check.url === url);

    return link?.status || "OK";
  };

  parent?.access?.accessUrls?.forEach((entry) => {
    const { proxyUrl, loginRequired } = getProxyUrl(
      entry.url || "",
      context?.user
    );

    res.push({
      __typename: "AccessUrl",
      origin: parseOnlineUrlToOrigin(entry.url),
      url: proxyUrl,
      loginRequired,
      note: entry.note,
      type: entry.type,
      status: linkStatus(proxyUrl) || "OK",
    });
  });

  if (parent?.access?.dbcWebArchive) {
    const archives = await context.datasources
      .getLoader("moreinfoWebarchive")
      .load(parent.pid);

    archives.forEach((archive) => {
      if (archive.url) {
        res.push({
          __typename: "AccessUrl",
          url: archive.url,
          origin: parseOnlineUrlToOrigin(archive.url),
          loginRequired: false,
          type: "RESOURCE",
          status: linkStatus(archive.url) || "OK",
        });
      }
    });
  }

  parent?.access?.ereol?.forEach((entry) => {
    res.push({
      __typename: "Ereol",
      origin: entry.origin,
      url: entry.url,
      canAlwaysBeLoaned: entry.canAlwaysBeLoaned,
      note: entry.note,
    });
  });

  if (parent?.access?.infomediaService?.id) {
    if (
      !["politiken", "jyllands-posten"].some((publication) =>
        parent?.hostPublication?.title?.toLowerCase()?.includes(publication)
      )
    ) {
      res.push({
        __typename: "InfomediaService",
        id: parent.access.infomediaService.id,
      });
    }

    // If this is a omitted online access for politiken or jyllands-posten.
    else {
      const hasPhysical = parent?.accessTypes?.some(
        (t) => (t?.code ?? "").toUpperCase() === "PHYSICAL"
      );

      if (hasPhysical) {
        parent = {
          ...parent,
          access: {
            ...(parent.access ?? {}),
            interLibraryLoanIsPossible: true,
          },
        };
      }
    }
  }

  // Get the issn for this article or periodica
  const issn =
    parent?.access?.digitalArticleService?.issn ||
    parent?.identifiers?.find?.((id) => id.type === "ISSN")?.value;

  if (issn) {
    const journals = await context.datasources
      .getLoader("statsbiblioteketJournals")
      .load("");
    const articleissn = issn.replace(/[^a-z\d]/gi, "");
    const hasJournal = journals && journals[articleissn];
    if (hasJournal) {
      res.push({
        __typename: "DigitalArticleService",
        issn: articleissn,
      });
    }
  }

  if (parent?.access?.interLibraryLoanIsPossible) {
    //if collectionidentifiers includes 870970-accessnew it may be loaned
    const isNewMaterial =
      parent?.collectionIdentifiers?.includes("870970-accessnew");

    const forLoan = checkInterLibraryLoan(parent, context);
    res.push({
      __typename: "InterLibraryLoan",
      loanIsPossible: forLoan,
      accessNew: isNewMaterial,
    });
  }

  // Return array containing all types of access
  return _sortOnlineAccess(res);
}

/**
 * municipality number is the second|third|fourth digit in agencyId
 * @param agencyId
 * @returns {string | undefined}
 */
function parseForMunicipalityNumber(agencyId) {
  return agencyId?.substring(1, 4);
}

/**
 * This one is for ebook.plus - we need to go via a proxy url (if user is logged in)
 * @param url
 * @param user
 * @returns {*}
 */
export function getProxyUrl(url, user) {
  const municipality =
    user?.municipality ||
    parseForMunicipalityNumber(user?.municipalityAgencyId);

  // check if we should proxy this url - for now it is ebookcentral and ebscohost
  const proxyMe =
    url.indexOf("ebookcentral") !== -1 || url.indexOf("ebscohost") !== -1;
  if (proxyMe) {
    // check if user is logged in
    if (user?.userId) {
      const realUrl = `https://bib${municipality}.bibbaser.dk/login?url=${url}`;
      return { proxyUrl: realUrl, loginRequired: proxyMe };
    }
  }

  return { proxyUrl: url, loginRequired: proxyMe };
}

/**
 * Some sources are not for loan (fagbibliografier) - for now a statis list:
 * This method is only called if interLibraryLoan is true - We know that for a fact
 * 159080-fagbib - Besættelsesbibliografien
 * 159081-fagbib - Bibliografi over Dansk Kunst
 * 159082-fagbib - Dansk Historisk Bibliografi
 * 159083-fagbib - Dansk Musiklitterær Bibliografi
 * 159084-fagbib - Kongelige Teater programartikler
 * 159085-fagbib - Dania Polyglotta
 * 159086-fagbib - Sportline
 * @param parent
 * @param context
 */
function checkInterLibraryLoan(parent, context) {
  // This is a blacklist of bases NOT to be loaned
  const notForLoanList = [
    "159080-fagbib",
    "159081-fagbib",
    "159082-fagbib",
    "159083-fagbib",
    "159084-fagbib",
    "159085-fagbib",
    "159086-fagbib",
  ];
  // we check on objectId - first part is the source
  const source = parent?.objectId?.split(":")[0];
  return !notForLoanList.includes(source);
}

/**
 * Get domain from url
 * @param url
 * @return String
 *  A parsed string eg. "DBC Webarkiv" or name of host eg. "infolink2003.elbo.dk"
 *
 * */
export function parseOnlineUrlToOrigin(url) {
  try {
    const parsedUrl = new URL(url);
    if (parsedUrl["host"] === "moreinfo.addi.dk") {
      return "DBC Webarkiv";
    } else {
      return (parsedUrl["host"] && parsedUrl["host"]) || "";
    }
  } catch (e) {
    return "";
  }
}

/**
 * Handle special cases - for now we want filmstriben/fjernleje BEFORE filmstriben/biblioteket AND
 * we want dansk film institut (dfi) to come last - it is not a 'real' url but a link to a description
 * @param onlineAccess
 * @return {*}
 * @private
 */
function _sortOnlineAccess(onlineAccess) {
  const specialSort = (a, b) => {
    // fjernleje should be on top
    if (b.url && b.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return 1;
    } else if (a.url && a.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return -1;
      // dfi is not a 'real' online url - sort low
    } else if (b.url && b.url.indexOf("dfi.dk") !== -1) {
      return -1;
    } else if (a.url && a.url.indexOf("dfi.dk") !== -1) {
      return 1;
    }
    return 0;
  };

  return onlineAccess.sort(specialSort);
}
