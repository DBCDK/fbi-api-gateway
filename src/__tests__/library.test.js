import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";

test("library - get branches for agency", async () => {
  const result = await performTestQuery({
    query: `
          query{
            branches(agencyid: "710100"){
              hitcount
              result {
                borrowerCheck
                agencyId
                branchId
                name
                openingHours
                userParameters {
                  userParameterType
                  parameterRequired
                }
                postalAddress
                postalCode
                city
                pickupAllowed
                digitalCopyAccess
              }
            }
          }
        `,
    variables: {},
    context: { datasources: createMockedDataLoaders() },
  });
  expect(result).toMatchSnapshot();
});

test("library - get all", async () => {
  const result = await performTestQuery({
    query: `
        query{
            branches{
              hitcount
              result {
                borrowerCheck
                agencyName
                agencyId
                branchId
                name
                openingHours
                userParameters {
                  userParameterType
                  parameterRequired
                }
                postalAddress
                postalCode
                city
                pickupAllowed
                digitalCopyAccess
              }
            }
          }
        `,
    variables: { bibdkExcludeBranches: true },
    context: { datasources: createMockedDataLoaders() },
  });
  expect(result).toMatchSnapshot();
});
