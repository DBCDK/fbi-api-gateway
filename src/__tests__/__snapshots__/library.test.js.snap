// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`library - get all 1`] = `
Object {
  "data": Object {
    "branches": Object {
      "hitcount": 6541,
      "result": Array [
        Object {
          "agencyId": "800010",
          "agencyName": "Det Kgl. Bibliotek",
          "borrowerCheck": true,
          "branchId": "800044",
          "city": "Aarhus C",
          "digitalCopyAccess": false,
          "name": "AU Bartholins Allé, Aarhus",
          "openingHours": null,
          "pickupAllowed": true,
          "postalAddress": "Bartholins Allé 11",
          "postalCode": "8000",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "800010",
          "agencyName": "Det Kgl. Bibliotek",
          "borrowerCheck": true,
          "branchId": "800041",
          "city": "Herning",
          "digitalCopyAccess": false,
          "name": "AU Birk Centerpark, Herning",
          "openingHours": null,
          "pickupAllowed": true,
          "postalAddress": "Birk Centerpark 15",
          "postalCode": "7400",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "800010",
          "agencyName": "Det Kgl. Bibliotek",
          "borrowerCheck": true,
          "branchId": "800039",
          "city": "København NV",
          "digitalCopyAccess": false,
          "name": "AU Emdrup, København",
          "openingHours": null,
          "pickupAllowed": true,
          "postalAddress": "Tuborgvvej 164",
          "postalCode": "2400",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "800010",
          "agencyName": "Det Kgl. Bibliotek",
          "borrowerCheck": true,
          "branchId": "800038",
          "city": "Aarhus V",
          "digitalCopyAccess": false,
          "name": "AU Fuglesangs Allé, Aarhus",
          "openingHours": null,
          "pickupAllowed": true,
          "postalAddress": "Fuglesangs Alle 4",
          "postalCode": "8210",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "800010",
          "agencyName": "Det Kgl. Bibliotek",
          "borrowerCheck": true,
          "branchId": "800037",
          "city": "Aarhus C",
          "digitalCopyAccess": false,
          "name": "AU Kasernen, Aarhus",
          "openingHours": null,
          "pickupAllowed": true,
          "postalAddress": "Langelandsgade 151",
          "postalCode": "8000",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "800010",
          "agencyName": "Det Kgl. Bibliotek",
          "borrowerCheck": true,
          "branchId": "800042",
          "city": "Aarhus N",
          "digitalCopyAccess": false,
          "name": "AU Katrinebjerg, Aarhus",
          "openingHours": null,
          "pickupAllowed": true,
          "postalAddress": "Finlandsgade 21, 1. sal",
          "postalCode": "8200",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "852130",
          "agencyName": "AU Library, Kalø",
          "borrowerCheck": true,
          "branchId": "852130",
          "city": "Rønde",
          "digitalCopyAccess": false,
          "name": "AU Library, Kalø",
          "openingHours": "8-16",
          "pickupAllowed": true,
          "postalAddress": "Grenåvej 14
Kalø",
          "postalCode": "8410",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "831040",
          "agencyName": "AU Library Katrinebjerg",
          "borrowerCheck": true,
          "branchId": "872120",
          "city": "Aarhus N",
          "digitalCopyAccess": false,
          "name": "AU Library, Katrinebjerg",
          "openingHours": "Mandag-fredag kl. 10.00-15.00",
          "pickupAllowed": true,
          "postalAddress": "Helsingforsgade 12 Bygning 5335 - 1. sal",
          "postalCode": "8200",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "820010",
          "agencyName": "Det Kgl. Bibliotek - Aarhus",
          "borrowerCheck": true,
          "branchId": "850920",
          "city": "Risskov",
          "digitalCopyAccess": false,
          "name": "AU Library, Psykiatri",
          "openingHours": "Mandag kl. 9-15
Tirsdag  kl. 9-15
Onsdag kl. 12-15
Torsdag kl. 9-15
Fredag   kl. 9-15",
          "pickupAllowed": true,
          "postalAddress": "Skovagervej 2, Indg. 34,  1. sal tv.",
          "postalCode": "8240",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "850800",
          "agencyName": "AU Library, Silkeborg",
          "borrowerCheck": true,
          "branchId": "850801",
          "city": "Roskilde",
          "digitalCopyAccess": false,
          "name": "AU Library, Roskilde",
          "openingHours": "Henvendelse via e-mail: bibldmu@dmu.dk",
          "pickupAllowed": true,
          "postalAddress": "Frederiksborgvej 399",
          "postalCode": "4000",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
      ],
    },
  },
}
`;

exports[`library - get branches for agency 1`] = `
Object {
  "data": Object {
    "branches": Object {
      "hitcount": 25,
      "result": Array [
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710110",
          "city": "København NV",
          "digitalCopyAccess": false,
          "name": "Biblioteket Rentemestervej",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Rentemestervej 76",
          "postalCode": "2400",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710114",
          "city": "København S",
          "digitalCopyAccess": false,
          "name": "Bibliotekshuset Rodosvej",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Rodosvej 4",
          "postalCode": "2300",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710104",
          "city": "København N",
          "digitalCopyAccess": false,
          "name": "Blågårdens Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker",
          "pickupAllowed": true,
          "postalAddress": "Blågårds Plads 5",
          "postalCode": "2200",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710105",
          "city": "Brønshøj",
          "digitalCopyAccess": false,
          "name": "Brønshøj Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Krabbesholmvej 3
",
          "postalCode": "2700",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710106",
          "city": "København K",
          "digitalCopyAccess": false,
          "name": "Christianshavns Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Dronningensgade 53",
          "postalCode": "1420",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710100",
          "city": "København K",
          "digitalCopyAccess": false,
          "name": "Hovedbiblioteket, Krystalgade",
          "openingHours": "Åbningstider
Se åbningstider på:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          "pickupAllowed": true,
          "postalAddress": "Krystalgade 15",
          "postalCode": "1172",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710107",
          "city": "Brønshøj",
          "digitalCopyAccess": false,
          "name": "Husum Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Frederikssundsvej 290",
          "postalCode": "2700",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710108",
          "city": "København S",
          "digitalCopyAccess": false,
          "name": "Islands Brygge Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Njalsgade 15",
          "postalCode": "2300",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710118",
          "city": "Vanløse",
          "digitalCopyAccess": false,
          "name": "Kulturstationen Vanløse. Biblioteket",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Jernbane Alle 38",
          "postalCode": "2720",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "borrowerCheck": true,
          "branchId": "710111",
          "city": "København N",
          "digitalCopyAccess": false,
          "name": "Nørrebro Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Nørrebrohallen
Nørrebrogade 208",
          "postalCode": "2200",
          "userParameters": Array [
            Object {
              "parameterRequired": true,
              "userParameterType": "USERID",
            },
            Object {
              "parameterRequired": false,
              "userParameterType": "USERMAIL",
            },
          ],
        },
      ],
    },
  },
}
`;
